room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/InvalidationTrackerBehavioralTest.java

```diff
@@ -0,0 +1,254 @@
+                                    if (localLatch == null) {
+                                        // Spurious invalidation callback; this might occur due to a
+                                        // large delay beyond the provisioned margin, or due to a
+                                        // bug in the code under test
+                                        spuriousInvalidations.incrementAndGet();
+                                    } else {
+                                        localLatch.countDown();
+                                    }
+                                }
+                            }
+                        }
+                );
+
+                try {
+                    // Resets latch and updates missedInvalidations when change notification failed
+                    for (int i = 0; i < iterations; ++i) {
+                        // The Counter table exists just to make InvalidationTracker's life more
+                        // difficult, we are not interested in notifications from this one;
+                        // inserts may trigger undefined invalidation callback behavior,
+                        // depending on table update timing
+                        db.counter1Dao().insert(new Counter1());
+
+                        // Use variable delay to detect different kinds of timing-related problems
+                        Thread.sleep(delayMillis, delayNanos);
+
+                        db.runInTransaction(() -> {
+                            db.counter2Dao().insert(new Counter2());
+
+                            // Flag that we have inserted a new value, expect invalidation callback;
+                            // do this as late as possible prior to the end of the transaction;
+                            // this might cause an occasional false negative due to a race,
+                            // where a buggy InvalidationTracker could log successful tracking
+                            // even though the transaction is not completed yet, but it does not
+                            // matter much, as this is an intentionally flaky test; on another run
+                            // it should become apparent that InvalidationTracker is buggy.
+                            latch = new CountDownLatch(1);
+                        });
+
+                        // Use sufficient delay to give invalidation tracker ample time to catch up;
+                        // this would need to be increased if the test has false positives.
+                        if (!latch.await(10L, TimeUnit.SECONDS)) {
+                            // The tracker still has not been called, log an error
+                            missedInvalidations.incrementAndGet();
+                        }
+
+                        latch = null;
+                    }
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
```

Comment:
If `InterruptedException ` will just throw then move it to just wrap `sleep` and `await`.